File: Dataset/OOP/AnimalAbstract.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package AnimalAbstract;

import Animal.*;

/**
 *
 * @author GIA KINH
 */
abstract public class Animal {
    public abstract void greeting();
    
}

public class Dog extends Animal {

    public Dog() {
    }

    @Override
    public String toString() {
        return "Dog[" + super.toString() + ']';
    }

    @Override
    public void greeting() {
        System.out.println("Woof!");
    }

    public void greeting(Dog another) {
        System.out.println("Woooooooooof!");
    }

}

public class BigDog extends Dog {

    public BigDog() {
    }

    @Override
    public String toString() {
        return "BigDog[" + super.toString() + ']';
    }

    @Override
    public void greeting() {
        System.out.println("Woow!");
    }

    @Override
    public void greeting(Dog another) {
        System.out.println("Woooooowwwww!");
    }
}


public class Cat extends Animal {

    public Cat() {
    }

    @Override
    public String toString() {
        return "Cat[" + super.toString() + ']';
    }

    @Override
    public void greeting() {
        System.out.println("Meow!");
    }

}


/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package AnimalAbstract;

import Animal.*;

/**
 *
 * @author GIA KINH
 */
public class TestMain {

    public static void main(String[] args) {
        // Using the subclasses
        Cat cat1 = new Cat();
        cat1.greeting();
        Dog dog1 = new Dog();
        dog1.greeting();
        BigDog bigDog1 = new BigDog();
        bigDog1.greeting();

        // Using Polymorphism
        Animal animal1 = new Cat();
        animal1.greeting();
        Animal animal2 = new Dog();
        animal2.greeting();
        Animal animal3 = new BigDog();
        animal3.greeting();
        // Animal animal4 = new Animal();   // Error!!! Animal is abstract; cannot be instantiated !

        // Downcast
        Dog dog2 = (Dog)animal2;
        BigDog bigDog2 = (BigDog)animal3;
        Dog dog3 = (Dog)animal3;
        // Cat cat2 = (Cat)animal2;        // Error!!! Dog cannot be cast to Cat !

        dog2.greeting(dog3);
        dog3.greeting(dog2);
        dog2.greeting(bigDog2);
        bigDog2.greeting(dog2);
        bigDog2.greeting(bigDog1);

        bigDog1.greeting(bigDog2);
    }
}

File: Dataset/Advanced/Base/OOP/Python/AnimalAbstract.py
from abc import ABC, abstractmethod

class Animal(ABC):
    @abstractmethod
    def greeting(self):
        pass

class Dog(Animal):
    def __init__(self):
        pass

    def __str__(self):
        return f"Dog[{super().__str__()}]"

    def greeting(self):
        print("Woof!")

    def greeting_dog(self, another):
        print("Woooooooooof!")

class BigDog(Dog):
    def __init__(self):
        pass

    def __str__(self):
        return f"BigDog[{super().__str__()}]"

    def greeting(self):
        print("Woow!")

    def greeting_dog(self, another):
        print("Woooooowwwww!")

class Cat(Animal):
    def __init__(self):
        pass

    def __str__(self):
        return f"Cat[{super().__str__()}]"

    def greeting(self):
        print("Meow!")

def main():
    # Using the subclasses
    cat1 = Cat()
    cat1.greeting()
    dog1 = Dog()
    dog1.greeting()
    bigDog1 = BigDog()
    bigDog1.greeting()

    # Using Polymorphism
    animal1 = Cat()
    animal1.greeting()
    animal2 = Dog()
    animal2.greeting()
    animal3 = BigDog()
    animal3.greeting()
    # animal4 = Animal()   # Error!!! Animal is abstract; cannot be instantiated !

    # Downcast
    dog2 = animal2
    bigDog2 = animal3
    dog3 = animal3
    # cat2 = animal2        # Error!!! Dog cannot be cast to Cat !

    dog2.greeting_dog(dog3)
    dog3.greeting_dog(dog2)
    dog2.greeting_dog(bigDog2)
    bigDog2.greeting_dog(dog2)
    bigDog2.greeting_dog(bigDog1)

    bigDog1.greeting_dog(bigDog2)

if __name__ == "__main__":
    main()